sklearn/utils/validation.py

```diff
@@ -980,3 +980,60 @@ def check_scalar(x, name, target_type, min_val=None, max_val=None):
 
     if max_val is not None and x > max_val:
         raise ValueError('`{}`= {}, must be <= {}.'.format(name, x, max_val))
+
+
+def _check_sample_weight(sample_weight, y=None, n_samples=None, dtype=None,
+                         order=None, **kwargs):
+    """Validate sample weights
+
+    Parameters
+    ----------
+    sample_weight : {ndarray, Number or None}
+       input sample weights
+    y: ndarray or None
+       target variable. Either y or n_samples must be provided.
+    n_samples: int or None
+       expected number of samples. Either y or n_samples must be provided.
+    dtype: dtype
+       dtype of the validated sample_weight
+    order : 'F', 'C' or None (default=None)
+        Whether an array will be forced to be fortran or c-style.
+        When order is None (default), if ``sample_weights`` is an ndarray,
+        nothing is ensured about the memory layout of the output array,
+        otherwise it will be of 'C' order by default.
+
+    kwargs:
+       additional parameters to pass to check_array
+
+    Parameters
+    ----------
+    sample_weight : ndarray, shape=(n_samples,)
+       validated sample weights
+    """
+    if n_samples is not None and y is not None:
+        raise ValueError('Only one of y, n_samples must be provided!')
+    elif y is not None:
+        n_samples = y.shape[0]
+
+    if sample_weight is None or isinstance(sample_weight, numbers.Number):
+        if order is None:
+            order = 'C'
+        if sample_weight is None:
+            sample_weight = np.ones(n_samples, dtype=dtype, order=order)
+        else:
+            sample_weight = np.full(n_samples, sample_weight,
+                                    dtype=dtype, order=order)
+    else:
+        if dtype is None:
+            dtype = [np.float64, np.float32]
+        sample_weight = check_array(
```

Comment:
Are there cases where we don't want `sample_weight.dtype` to be the same as `X.dtype` ?
Because if there aren't any we can pass X as a parameter, force X.dtype and get rid of the `n_samples` parameter.

Reply:
> Are there cases where we don't want sample_weight.dtype to be the same as X.dtype ?

That's a very good question, but I'm not 100% sure where `sample_weight` are used across estimators. For instance, take the case or some regressor where `X` is `float32`, and `y` is `float64`. Right now `sample_weight` would be `float64`  by default. Would it make sense to also make it float32? I don't know, maybe. Would it break something, and do we currently have sufficient tests to detect it? I don't know either.

So here I went with the safer road of just adding this refactoring without changing dtypes. The second part could be added but I think it would deserve a more careful look as part of https://github.com/scikit-learn/scikit-learn/issues/11000 etc.